go.mod

```diff
@@ -1,6 +1,6 @@
 module github.com/opencontainers/runc
 
-go 1.22
+go 1.22.4
```

Comment:
I don't know whether there are some projects which require runc lib, but they don't use `nsexec`, and can't upgrade to 1.22.4 or higher version.
These projects will compile failure?

Reply:
Could we use a per file version constraint instead so the parts of runc that do need 1.22+ fail to build without it but everything else can still be used?

Reply:
It seems like golang doesn't support minimal version grammer in the go compile directives.

Reply:
Ah, we could specify `//go:build go1.22` (which means 1.22+) but _not_ `go1.22.4`, I see :disappointed:

Reply:
I might be missing something, but I can't see why this is a problem. For several reasons:

1. We are already specifying go 1.22, so go 1.22.0 is enforced as the min go version (unless you are using a very old version of go, in which case go 1.22 or 1.22.4 doesn't change anything)
1. Go < 1.22 is not supported upstream. I don't understand why we should worry about that
1. I'm quite sure we are using features of go 1.22 or 1.21, so some parts of the code won't build with older versions of go (besides the nsexec issue)

Am I missing something?

Reply:
> Am I missing something?

https://github.com/cri-o/cri-o/blob/main/go.mod

Reply:
What I mean is that, there are some projects maybe only work on go version `less than 1.22.4`.

Reply:
Hmm, that doesn't mean that, right? That just means the minimum go version is lower, right?

Reply:
The `toolchain` in `go.mod` will fix golang to a specific version, for example `1.22.2` in `cri-o`, but it will cause an error when doing `go mod vendor`:
```
go: module ../../opencontainers/runc requires go >= 1.22.4 (running go 1.22.2)
```

Reply:
So they will have to update the minimum version to 1.22.4, right? What is the issue? I don't really understand.

Reply:
I don't know why they didn't update the minimum version of go, maybe there are some issues in their project, I don't know.
Before this patch, they can compile in 1.22.2 successful, so you mean it's the job of their own, if they want to bump runc to 1.2.0-*, they need to update the min go version. We should not consider such type projects which require runc (without nsexec).

Reply:
And that is only if they don't merge this PR before upgrading the runc module: https://github.com/cri-o/cri-o/pull/8597

The runc 1.2.0 final release won't be finished in the next few days.

Reply:
@lifubang I don't understand where that would be a problem. Sure, you need a newer version of go, instead of 1.22.2, you need 1.22.4. When that is an issue that we really want to let users compile this with 1.22 < 1.22.4?

I mean, 1.22.4 is out for some time (1.22.7 is the latest now), if it introduced a regression, they didn't report it and then all the go 1.22.x > 2 cause an issue for them... Sure, this will be problematic. If they reported the regression, it seems likely that by 1.22.7 it was already fixed.

I think it is unlikely, and we can revisit this decision if some user is actually in such a weird situation.

Reply:
Sorry, taking a step back I think I see what you mean. It's just about projects using "1.22" in the go line that will need to change it when upgrading runc, and maybe that change is not even needed (as they are not using nsexec).

Luckily, that is why the go team added the "toolchain" line with this min go version change. We can _suggest_  go 1.22.4 (using toolchain 1.22.4) when compiling runc, but for projects that use it as a module the go line applies.

I'll update the PR to do that later. I'll mark it as draft until then.

Thanks for the feedback :)

Reply:
I think this is solved in the last push, feel free to re-open if I'm missing something